File: parser/parser.go
package parser

import (
	"fmt"
	"monkey/ast"
	"monkey/lexer"
	"monkey/token"
	"strconv"
)

// Order of operator precedence
const (
	LOWEST      = 1
	EQUALS      = 2 // ==
	LESSGREATER = 3 // > or <
	SUM         = 4 // +
	PRODUCT     = 5 // *
	PREFIX      = 6 // -X, --X, !X, !!X, &X, *X
	CALL        = 7 // myFunction(X)
)

type (
	prefixParseFn func() ast.Expression
	infixParseFn  func(ast.Expression) ast.Expression // arg is “left side” of the infix operator that’s being parsed
)

// Associates token types with their precedence
var precedences = map[token.TokenType]int{
	token.EQ:       EQUALS,
	token.NOT_EQ:   EQUALS,
	token.LT:       LESSGREATER,
	token.GT:       LESSGREATER,
	token.PLUS:     SUM,
	token.MINUS:    SUM,
	token.SLASH:    PRODUCT,
	token.ASTERISK: PRODUCT,
}

type Parser struct {
	lexer  *lexer.Lexer
	errors []string

	curToken  token.Token
	peekToken token.Token // Token after curToken

	prefixParseFns map[token.TokenType]prefixParseFn
	infixParseFns  map[token.TokenType]infixParseFn
}

func New(lexer *lexer.Lexer) *Parser {
	parser := &Parser{
		lexer:  lexer,
		errors: []string{},
	}

	parser.prefixParseFns = make(map[token.TokenType]prefixParseFn)
	parser.registerPrefix(token.IDENT, parser.parserIdentifier)
	parser.registerPrefix(token.INT, parser.parseIntegerLiteral)
	parser.registerPrefix(token.BANG, parser.parsePrefixExpression)
	parser.registerPrefix(token.MINUS, parser.parsePrefixExpression)

	// Register all possible "infix operators"
	parser.infixParseFns = make(map[token.TokenType]infixParseFn)
	parser.registerInfix(token.PLUS, parser.parseInfixExpression)
	parser.registerInfix(token.MINUS, parser.parseInfixExpression)
	parser.registerInfix(token.SLASH, parser.parseInfixExpression)
	parser.registerInfix(token.ASTERISK, parser.parseInfixExpression)
	parser.registerInfix(token.EQ, parser.parseInfixExpression)
	parser.registerInfix(token.NOT_EQ, parser.parseInfixExpression)
	parser.registerInfix(token.LT, parser.parseInfixExpression)
	parser.registerInfix(token.GT, parser.parseInfixExpression)

	// Read two tokens, so curToken and peekToken are both set
	parser.nextToken()
	parser.nextToken()

	return parser
}

func (parser *Parser) parserIdentifier() ast.Expression {
	return &ast.Identifier{Token: parser.curToken, Value: parser.curToken.Literal}
}

/*
Construct an *ast.InfixExpression node, with left being in the `Left` field.
Assigns the precedence of the current token (which is the operator of the infix expression)
to the local variable precedence, before advancing the tokens by calling `nextToken`
and filling the `Right` field of the node with another call to
`parseExpression` - this time passing in the precedence of the operator token.
*/
func (parser *Parser) parseInfixExpression(left ast.Expression) ast.Expression {
	expression := &ast.InfixExpression{
		Token:    parser.curToken,
		Operator: parser.curToken.Literal,
		Left:     left,
	}

	precedence := parser.curPrecedence()
	parser.nextToken()
	expression.Right = parser.parseExpression(precedence)

	return expression
}

func (parser *Parser) nextToken() {
	parser.curToken = parser.peekToken
	parser.peekToken = parser.lexer.NextToken()
}

func (parser *Parser) parseStatement() ast.Statement {
	switch parser.curToken.Type {
	case token.LET:
		return parser.parseLetStatement()
	case token.RETURN:
		return parser.parseReturnStatement()
	default:
		return parser.parseExpressionStatement()
	}
}

func (parser *Parser) parseLetStatement() *ast.LetStatement {
	statement := &ast.LetStatement{Token: parser.curToken}

	if !parser.expectPeek(token.IDENT) {
		return nil
	}

	statement.Name = &ast.Identifier{
		Token: parser.curToken,
		Value: parser.curToken.Literal,
	}

	if !parser.expectPeek(token.ASSIGN) {
		return nil
	}
	// TODO: We're skipping the expressions until we
	// encounter a semicolon
	for !parser.curTokenIs(token.SEMICOLON) {
		parser.nextToken()
	}

	return statement
}

func (parser *Parser) parseReturnStatement() *ast.ReturnStatement {
	statement := &ast.ReturnStatement{Token: parser.curToken}

	parser.nextToken()
	// TODO: We're skipping the expressions until we
	// encounter a semicolon
	for !parser.curTokenIs(token.SEMICOLON) {
		parser.nextToken()
	}
	return statement
}

func (parser *Parser) parseExpressionStatement() *ast.ExpressionStatement {
	statement := &ast.ExpressionStatement{Token: parser.curToken}

	statement.Expression = parser.parseExpression(LOWEST)

	if parser.peekTokenIs(token.SEMICOLON) {
		parser.nextToken()
	}

	return statement
}

// Checking whether we have a parsing function associated
// with `parser.curToken.Type` in the prefix position
func (parser *Parser) parseExpression(precedence int) ast.Expression {
	prefix := parser.prefixParseFns[parser.curToken.Type]

	if prefix == nil {
		parser.noPrefixParseFnError(parser.curToken.Type)
		return nil
	}

	leftExp := prefix()
	for !parser.peekTokenIs(token.SEMICOLON) && precedence < parser.peekPrecedence() {
		infix := parser.infixParseFns[parser.peekToken.Type]

		if infix == nil {
			return leftExp
		}

		parser.nextToken()
		leftExp = infix(leftExp)
	}

	return leftExp
}

func (parser *Parser) parsePrefixExpression() ast.Expression {
	expression := &ast.PrefixExpression{
		Token:    parser.curToken,
		Operator: parser.curToken.Literal,
	}

	parser.nextToken()
	expression.Right = parser.parseExpression(PREFIX)

	return expression
}

func (parser *Parser) curTokenIs(tok token.TokenType) bool {
	return parser.curToken.Type == tok
}

func (parser *Parser) peekTokenIs(tok token.TokenType) bool {
	return parser.peekToken.Type == tok
}

func (parser *Parser) expectPeek(tok token.TokenType) bool {
	if parser.peekTokenIs(tok) {
		parser.nextToken()
		return true
	}

	parser.peekError(tok)
	return false
}

/*
	Constructs the root node of the AST, an *ast.Program.
	It then iterates over every token in the input until it
	encounters an token.EOF token. It does this by repeatedly
	calling nextToken, which advances both parser.curToken
	and parser.peekToken.
*/
func (parser *Parser) ParseProgram() *ast.Program {
	program := &ast.Program{}
	program.Statements = []ast.Statement{}

	for parser.curToken.Type != token.EOF {
		statement := parser.parseStatement()

		if statement != nil {
			program.Statements = append(program.Statements, statement)
		}

		parser.nextToken()
	}

	return program
}

func (parser *Parser) Errors() []string {
	return parser.errors
}

func (parser *Parser) peekError(t token.TokenType) {
	msg := fmt.Sprintf("expected next token to be %s, got %s instead",
		t, parser.peekToken.Type)
	parser.errors = append(parser.errors, msg)
}

func (parser *Parser) registerPrefix(tokenType token.TokenType, fn prefixParseFn) {
	parser.prefixParseFns[tokenType] = fn
}

func (parser *Parser) registerInfix(tokenType token.TokenType, fn infixParseFn) {
	parser.infixParseFns[tokenType] = fn
}

func (parser *Parser) parseIntegerLiteral() ast.Expression {
	literal := &ast.IntegerLiteral{Token: parser.curToken}
	value, err := strconv.ParseInt(parser.curToken.Literal, 0, 64)

	if err != nil {
		msg := fmt.Sprintf("could not parse %q as integer", parser.curToken.Literal)
		parser.errors = append(parser.errors, msg)
		return nil
	}

	literal.Value = value
	return literal
}

func (parser *Parser) noPrefixParseFnError(t token.TokenType) {
	msg := fmt.Sprintf("no prefix parse function for %s found", t)
	parser.errors = append(parser.errors, msg)
}

// Returns the precedence associated with the token type of peekToken.
// Returns the lowest precedence if one cannot be found.
func (parser *Parser) peekPrecedence() int {
	if parser, ok := precedences[parser.peekToken.Type]; ok {
		return parser
	}

	return LOWEST
}

// Returns the precedence associated with the token type of curToken.
// Returns the lowest precedence if one cannot be found.
func (parser *Parser) curPrecedence() int {
	if parser, ok := precedences[parser.curToken.Type]; ok {
		return parser
	}

	return LOWEST
}

File: lib/monkey/parser.rb
# typed: strict
# frozen_string_literal: true

module Monkey
  # rubocop:disable Metrics/ClassLength
  class Parser
    extend T::Sig

    LOWEST         = 0
    EQUALS         = 10 # ==
    LESS_GREATER   = 20 # > or <
    SUM_MINUS      = 30 # +
    PRODUCT_DIVIDE = 40 # * /
    PREFIX         = 50 # -X, --X, !X, !!X, &X, *X
    CALL           = 60 # myFunction(X)

    PRECEDENCES = T.let({
      Token::EQ       => EQUALS,
      Token::NOT_EQ   => EQUALS,
      Token::LT       => LESS_GREATER,
      Token::GT       => LESS_GREATER,
      Token::PLUS     => SUM_MINUS,
      Token::MINUS    => SUM_MINUS,
      Token::SLASH    => PRODUCT_DIVIDE,
      Token::ASTERISK => PRODUCT_DIVIDE
    }.freeze, T::Hash[String, Integer])

    sig { params(lexer: Lexer).void }
    def initialize(lexer)
      @lexer = lexer
      @errors        = T.let([], T::Array[String])
      # Read two tokens, so curToken and peekToken are both set
      @curr_token    = T.let(@lexer.next_token!, Token)
      @peek_token    = T.let(@lexer.next_token!, Token)
    end

    sig { returns(AST::Program) }
    def parse_program!
      program = AST::Program.new

      parse!(program) until @curr_token.eof?

      program
    end

    private

    sig { params(program: AST::Program).void }
    def parse!(program)
      statement = parse_statement
      program.statements << statement if statement

      next_token!
    end

    sig { returns(T.nilable(AST::Statement)) }
    def parse_statement
      case @curr_token.type
      when Token::LET then parse_let_statement!
      when Token::RETURN then parse_return_statement
      else
        parse_expression_statement
      end
    end

    sig { returns(T.nilable(AST::LetStatement)) }
    def parse_let_statement!
      token = @curr_token
      return nil unless expect_peek!(Token::IDENTIFIER)

      identifier = AST::Identifier.new @curr_token, @curr_token.literal

      return nil unless expect_peek!(Token::ASSIGN)

      next_token!
      expression = parse_expression(LOWEST)
      next_token! if peek_token_is?(Token::SEMICOLON)

      AST::LetStatement.new token, identifier, expression
    end

    sig { params(type: String).returns(T::Boolean) }
    def expect_peek!(type)
      if peek_token_is? type
        next_token!
        return true
      end

      peek_error type
      false
    end

    sig { params(type: String).returns(T::Boolean) }
    def peek_token_is?(type)
      @peek_token.type == type
    end

    sig { params(precedence: Integer).returns(AST::Expression) }
    def parse_expression(precedence); end

    sig { params(type: String).void }
    def peek_error(type); end

    sig { returns(AST::ReturnStatement) }
    def parse_return_statement; end

    sig { returns(AST::ExpressionStatement) }
    def parse_expression_statement; end

    sig { returns(AST::Identifier) }
    def parse_identifier
      AST::Identifier.new(
        @curr_token,
        @curr_token.literal ,
      )
    end

    sig { returns(AST::Identifier) }
    def parse_integer_literal
      AST::Identifier.new(
        @curr_token,
        @curr_token.literal
      )
    end

    sig { returns(AST::Identifier) }
    def parse_prefix_expression
      AST::Identifier.new(
        @curr_token,
        @curr_token.literal
      )
    end

    sig { returns(AST::Identifier) }
    def parse_boolean
      AST::Identifier.new(
        @curr_token,
        @curr_token.literal
      )
    end

    sig { returns(AST::Identifier) }
    def parse_grouped_expression
      AST::Identifier.new(
        @curr_token,
        @curr_token.literal
      )
    end

    sig { returns(AST::Identifier) }
    def parse_if_expression
      AST::Identifier.new(
        @curr_token,
        @curr_token.literal
      )
    end

    sig { void }
    def next_token!
      @curr_token = @peek_token
      @peek_token  = @lexer.next_token!
    end

    sig { params(left: AST::Expression).returns(AST::Expression) }
    def parse_infix_expression(left)
      TODO:
      expression = AST::InfixExpression.new
      # expression := &ast.InfixExpression{
      #   Token:    parser.curToken,
      #   Operator: parser.curToken.Literal,
      #   Left:     left,
      # }

      # precedence := parser.curPrecedence()
      # parser.nextToken()
      # expression.Right = parser.parseExpression(precedence)

      # return expression
    end

    PrefixDispatcher = T.let({
      Token::IDENTIFIER => instance_method(:parse_identifier),
      Token::INT        => instance_method(:parse_integer_literal),
      Token::BANG       => instance_method(:parse_prefix_expression),
      Token::MINUS      => instance_method(:parse_prefix_expression),
      Token::TRUE       => instance_method(:parse_boolean),
      Token::FALSE      => instance_method(:parse_boolean),
      Token::L_PAREN    => instance_method(:parse_grouped_expression),
      Token::IF         => instance_method(:parse_if_expression)
    }.freeze, T::Hash[String, UnboundMethod])

    # TODO: This is stupid. What's a better way to say "is this token an infix operator"?
    InfixDispater = T.let({
      # Token::PLUS     => instance_method(:parseInfixExpression),
      # Token::MINUS    => instance_method(:parseInfixExpression),
      # Token::SLASH    => instance_method(:parseInfixExpression),
      # Token::ASTERISK => instance_method(:parseInfixExpression),
      # Token::EQ       => instance_method(:parseInfixExpression),
      # Token::NOT_EQ   => instance_method(:parseInfixExpression),
      # Token::LT       => instance_method(:parseInfixExpression),
      # Token::GT       => instance_method(:parseInfixExpression)
    }.freeze, T::Hash[String, UnboundMethod])
  end
  # rubocop:enable Metrics/ClassLength
end
